segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/mocks/InMemoryStorage.java

```diff
@@ -195,6 +197,14 @@ public boolean supportsTruncation() {
         return false;
     }
 
+    @Override
+    public Iterator<SegmentProperties> listSegments() {
+        synchronized (this) {
+            Collection<StreamSegmentData> copyValues = this.streamSegments.values();
```

Comment:
This is not making a copy. You need to do `new ArrayList<>(this.streamSegments.values())`  inside the `synchronized` block, and then move the following line out of that block.